### ISSUE TITLE
Wrong namespace for generatice TypeProviders

### ISSUE THREAD

**Issue by thinkbeforecoding** [fsprojects/FSharp.TypeProviders.SDK#384]
### Description

The WsdlProvider is a generative provider due to the use of WCF. To implement Wsdl Choice elements, some properties must have an attribute referencing the Type of provided classes.

When generating the typeof<SomeProvidedType>, the namespace emitted is the generic one used in the TP, but the type is actually emitted in the user's code namespace. When used, the actual type cannot be found.

### Repro steps

1. Create a TP type provider
1. In the type generation, create a ProvidedTypeDefinition call MyObject in TpNS namespace
2. In a provided function, emit a typeof this type.
3. Write a sample that use the type provider in a Sample namespace

### Expected behavior

When running the sample, the type should be found in the Sample namespace

### Actual behavior

The typeof jitting will fail, indicating that the type TpNS.MyObject cannot be found. The class is actually in Sample.MyObject.

### Known workarounds

There is no known workaround.

Is there a way to get the actual namespace where the generated code will be compiled ? Is this information available in the TypeProvider ?

### Related information 

* All
* main
* .NET Runtime & CoreCLR at least

**Comment by thinkbeforecoding** [comment]
I checked the IL ecma spec, and it seems everywhere else, types are referenced by index in a table, but in the case of attributes, the attribute blob data contains the name of the attribute... Except that once injected in the destination, the inner namespace used to create ProvidedTypeDefinition isn't the actual type namespace. An error is raised as soon as you try to load the type as the attribute type is not found.

It doesn't seem that the destination namespace is accessible in the type provider so it seems not doable... Am I wrong?

**Comment by thinkbeforecoding** [comment]
so it works as long as you reference external types. but not when using provided types.

**Comment by dsyme** [comment]
Yes this difference in encoding for type references in attributes has been a bane to rewriting tools before, it is certainly the problem here

**Comment by dsyme** [comment]
I think it is the rewrite phase in the compiler for generated assemblies which is missing this. I'll take a look if its a simple fix though could take a while to get through the works even if it is
